### PR TITLE
[IMP] web_editor: floating bar and commandbar scss review

### PR DESCRIPTION
**Pull request by morm-odoo**
Prior to this PR, there were unnecessary scss for the floating toolbar and the powerbox in the web editor.

First commit also fixes some design issues like alignement for the icons in the list of the powerbox, box shadows, and the responsiveness of the floating toolbar on mobile.

In commit [1], the toolbar dropdown had the display static added, but it causes problems to the responsiveness.
2th commit put the display back to dynamic on the dropdown and so, fixes the dropdown not showing on mobile because of the overflow-x on auto. The dynamic state also allows the dropdown to change direction (dropdown to dropup) when there's not enough space for the content of the dropdown.

3th commit adds IsMobile

4th commit uses isMobile on the all editor toolbars (frontend and backend)
The layout of the toolbar, for mobile, is modified for a better behaviour and design.
Prior to this, it's was a horizontal scroll.

[1]: 459d4e27a860f3f0308ee0984de6f789baa65424

task-3087826

### CODE REVIEW

File: addons/web/static/src/core/notebook/notebook.scss [hunk 1]
```diff
@@ -17,7 +17,6 @@
             }
         }
     }
-
```
Comment (stefanorigano) [inline addons/web/static/src/core/notebook/notebook.scss:20]: Unnecessary change

File: addons/web/static/src/core/notebook/notebook.scss [hunk 2]
```diff
@@ -78,4 +77,12 @@
             }
         }
     }
+    @media only screen and (max-width: 767px) {
```
Comment (stefanorigano) [inline addons/web/static/src/core/notebook/notebook.scss:80]: Use[ bootstrap default mixins](https://getbootstrap.com/docs/5.1/layout/breakpoints/#max-width).
Also, missing empty-line before selector
```suggestion

    @media only screen and (max-width: 767px) {
```

File: addons/web_editor/static/src/js/editor/odoo-editor/src/style.scss [hunk 3]
```diff
@@ -266,33 +265,38 @@
...
         background-color: white;
-
+        box-shadow: none;
+        height: auto;
+        
         .btn {
             color: black;
+            padding: 3.5px 4px !important;
+        }
+        .dropdown-menu{
```
Comment (stefanorigano) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/style.scss:281]: missing empty line before selector, missing empty character before opening brachet
```suggestion

        .dropdown-menu {
```

File: addons/web_editor/static/src/scss/wysiwyg.scss [hunk 4]
```diff
@@ -56,7 +56,6 @@ $o-we-zindex: $o-we-overlay-zindex + 1 !default;
     button {
         outline: none;
     }
-
```
Comment (stefanorigano) [inline addons/web_editor/static/src/scss/wysiwyg.scss:59]: unnecessary change

File: addons/web_editor/static/src/scss/wysiwyg.scss [hunk 5]
```diff
@@ -642,13 +642,17 @@ img.o_we_selected_image {
 }
 
 .o_edit_menu_popover {
+    box-shadow: rgba(0, 0, 0 , 0.1) 0px 4px 12px, rgba(0, 0, 0, 0.05) 0px 0px 4px;
+    border-color: rgba(0,0,0,0.025);
```
Comment (stefanorigano) [inline addons/web_editor/static/src/scss/wysiwyg.scss:646]: `box-shadow` and  `border-color` should be placed after position and sizing rules
ref: https://www.odoo.com/documentation/16.0/contributing/development/coding_guidelines.html#properties-order

File: addons/web/static/src/core/notebook/notebook.scss [hunk 6]
```diff
@@ -78,4 +77,12 @@
...
+        .o_notebook_content .oe-toolbar {
+            position: sticky;
```
Comment (stefanorigano) [inline addons/web/static/src/core/notebook/notebook.scss:82]: You can use our custom mixin
https://github.com/odoo/odoo/commit/5f1d32f3fb89f40740b25491caafb9a5710d555a#diff-3a53cced7e488eec86f2184a7fa68d1065cb1e0130689d0ebba047acd74c6774R30

File: addons/web/static/src/core/notebook/notebook.scss [hunk 7]
```diff
@@ -78,4 +77,12 @@
...
+            top: 0;
+            margin: 0 var(--notebook-margin-x, -16px);
```
Comment (stefanorigano) [inline addons/web/static/src/core/notebook/notebook.scss:84]: Since we're playing with horizontal  margins only here, is better to keep the vertical ones untouched.
One line more, but less headaches in case of future adaptations.

 So:
```suggestion
            margin-left: var(--notebook-margin-x, -16px);
            margin-right: var(--notebook-margin-x, -16px);
```

...btw,  why we need  `16px` ?

File: addons/web_editor/static/src/js/editor/odoo-editor/src/style.scss [hunk 8]
```diff
@@ -266,33 +265,38 @@
...
-        position: relative;
+        overflow-x: auto;
         visibility: visible;
         width: 100%;
+        border-bottom: 1px solid $border-color;
         border-radius: 0;
         background-color: white;
-
+        box-shadow: none;
+        height: auto;
```
Comment (stefanorigano) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/style.scss:275]: here too, watchout for CSS declarations order ;)

File: addons/web_editor/static/src/scss/wysiwyg.scss [hunk 9]
```diff
@@ -405,6 +404,7 @@ body:not(.editor_has_snippets) {
 }
 @media only screen and (max-width: 767px) {
     .oe-toolbar {
+        position: static;
```
Comment (stefanorigano) [inline addons/web_editor/static/src/scss/wysiwyg.scss:407]: Add one line in the commit's message explaining why this was needed (forum?)

File: addons/web_editor/static/src/js/editor/odoo-editor/src/style.scss [hunk 10]
```diff
@@ -266,33 +265,38 @@
...
+        
         .btn {
             color: black;
+            padding: 3.5px 4px !important;
```
Comment (stefanorigano) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/style.scss:279]: Avoid sub-pixels since are rendered inconsistently across browsers/monitors/OS...

File: addons/web_editor/static/src/scss/wysiwyg.dark.scss [hunk 11]
```diff
@@ -15,3 +15,9 @@
 .o_link_dialog_preview {
     --o-link-dialog-preview-border: #{$border-width} solid #{$border-color};
 }
+.o_colorpicker_sections, .colorpicker .o_colorpicker_sections .o_we_color_btn {
```
Comment (stefanorigano) [inline addons/web_editor/static/src/scss/wysiwyg.dark.scss:18]: `.colorpicker .o_colorpicker_sections .o_we_color_btn` is a very specific selector, can't we reduce its specificity?

File: addons/web_editor/static/src/scss/wysiwyg.dark.scss [hunk 12]
```diff
@@ -15,3 +15,9 @@
...
+    --bg: #{$o-gray-200};
+}
+.o_custom_gradient_btn{
```
Comment (stefanorigano) [inline addons/web_editor/static/src/scss/wysiwyg.dark.scss:21]:
```suggestion

.o_custom_gradient_btn {
```

File: addons/web_editor/static/src/scss/wysiwyg.scss [hunk 13]
```diff
@@ -133,6 +132,11 @@ $o-we-zindex: $o-we-overlay-zindex + 1 !default;
         }
 
         .o_colorpicker_section {
+            &.o_custom_gradient_editor{
+                .o_custom_gradient_btn{
+                    border: 1px solid var(--o-we-toolbar-border);
+                }
+            }
```
Comment (stefanorigano) [inline addons/web_editor/static/src/scss/wysiwyg.scss:139]: Missing empty line
```suggestion
            }
            
```

File: addons/web_editor/static/src/scss/wysiwyg.scss [hunk 14]
```diff
@@ -267,7 +267,6 @@ $o-we-zindex: $o-we-overlay-zindex + 1 !default;
         @include o-caret-down();
         margin-left: .3em
     }
-
```
Comment (stefanorigano) [inline addons/web_editor/static/src/scss/wysiwyg.scss:270]: unnecessary change

File: addons/web_editor/static/src/scss/wysiwyg.scss [hunk 15]
```diff
@@ -182,16 +186,12 @@ $o-we-zindex: $o-we-overlay-zindex + 1 !default;
 
 // Floating Toolbar
 .oe-toolbar.oe-floating {
-    gap: 0 $o-we-sidebar-content-field-control-item-spacing * .7;
-    grid-auto-flow: column;
+
```
Comment (stefanorigano) [inline addons/web_editor/static/src/scss/wysiwyg.scss:189]: unnecessary new line

File: addons/web_editor/static/src/js/editor/odoo-editor/src/style.scss [hunk 16]
```diff
@@ -266,33 +265,38 @@
...
     background: var(--oePowerbox__wrapper-bg, $o-white);
     color: $o-main-text-color;
     max-height: 40vh;
     box-sizing: border-box;
-    max-width: 100%;
-    box-shadow: 0px 3px 18px rgba(0, 0, 0, .23);
-    border-radius: 4px;
-    overflow: hidden;
-    display: flex;
+    box-shadow: rgba(0, 0, 0 , 0.1) 0px 4px 12px, rgba(0, 0, 0, 0.05) 0px 0px 4px;
```
Comment (stefanorigano) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/style.scss:299]: can't we use a combination of these variables? 
https://github.com/twbs/bootstrap/blob/1a6fdfae6b/scss/_variables.scss#L538

**Review by stefanorigano — CHANGES_REQUESTED** [review]
Hello @morm-odoo , thanks.
Some minor remarks, mostly about conventions

File: addons/web_editor/static/src/js/editor/odoo-editor/src/style.scss (hunk 16 above)
Comment (morm-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/style.scss]: I replaced it by the $box-shadow variable but I think we should change that variable to make it less harsh.

File: addons/web_editor/static/src/scss/wysiwyg.scss [hunk 17]
```diff
@@ -648,7 +651,11 @@ img.o_we_selected_image {
     // Prevent the edited link from being deselected when clicking between
     // buttons in the popover
     user-select: none;
-
+    box-shadow: $box-shadow;
+    border-color: rgba(0,0,0,0.025);
+    .popover-arrow::before{
```
Comment (stefanorigano) [inline addons/web_editor/static/src/scss/wysiwyg.scss:656]: Missing empty line before the selector and an empty space before the bracket ;)
```suggestion

    .popover-arrow::before {
```

File: addons/web_editor/static/src/scss/wysiwyg.scss [hunk 18]
```diff
@@ -403,8 +404,10 @@ body:not(.editor_has_snippets) {
         z-index: 1056;
     }
 }
-@media only screen and (max-width: 767px) {
+
+@include media-breakpoint-down(md) {
     .oe-toolbar {
+        position: static;
```
Comment (stefanorigano) [inline addons/web_editor/static/src/scss/wysiwyg.scss:410]: Can't we use `position-static position-md-absolute` classes combination?

File: addons/web_editor/static/src/scss/wysiwyg.dark.scss [hunk 19]
```diff
@@ -15,3 +15,9 @@
 .o_link_dialog_preview {
     --o-link-dialog-preview-border: #{$border-width} solid #{$border-color};
 }
+.o_colorpicker_sections, .o_we_color_btn {
```
Comment (stefanorigano) [inline addons/web_editor/static/src/scss/wysiwyg.dark.scss:18]: Add an empty row before each selector ;)

**Review by stefanorigano — CHANGES_REQUESTED** [review]
Thanks @morm-odoo , some minor remarks in the code.
I'll check the visual-result later on

File: addons/web_editor/static/src/scss/wysiwyg.scss (hunk 18 above)
Comment (morm-odoo) [reply addons/web_editor/static/src/scss/wysiwyg.scss:410]: The problem if I do that, is that the toolbar won't be sticky on mobile (@include o-position-sticky($top: 0px);) anymore in modules with notebook like project.

Comment (stefanorigano) [reply addons/web_editor/static/src/scss/wysiwyg.scss:410]: yes, ok, thanks

**Comment by stefanorigano** [comment]
Hello @morm-odoo , what's the status? should I review? Thanks

**Comment by morm-odoo** [comment]
@stefanorigano Hello, I have to fix it on the sidebar of mass_mailing first

**Comment by stefanorigano** [comment]
@morm-odoo , update: it should be feasible for framework-editor to push a commit on top in order to expose `isMobile` in the `xml`template.  They will have a look :)

@dmo-odoo , here it is, thanks 👍

**Comment by stefanorigano** [comment]
@dmo-odoo  🚀💯

**Comment by dmo-odoo** [comment]
@morm-odoo I have added a commit that should make `isMobile` accessible from the toolbar template. To be honest, I didn't take the time to try it, so let me know if it doesn't work and I'll have a second look.

File: addons/web_editor/static/src/xml/editor.xml [hunk 20]
```diff
@@ -97,8 +99,9 @@
 
             <div id="justify" class="btn-group dropdown">
                 <button type="button" class="btn dropdown-toggle"
-                    data-bs-toggle="dropdown" tabindex="-1" data-bs-display="static" aria-expanded="false">
+                    data-bs-toggle="dropdown" tabindex="-1" data-bs-original-title="Paragraph" aria-expanded="false">
                     <i id="paragraphDropdownButton" class="fa fa-align-left fa-fw" title="Text align"></i>
+                    <i id="paragraphDropdownButton" class="fa fa-align-left fa-fw"></i>
```
Comment (Zinston) [inline addons/web_editor/static/src/xml/editor.xml]: You remove this line in the next commit. I'm assuming this is a rebasing error :-)

File: addons/web/static/src/core/notebook/notebook.scss [hunk 21]
```diff
@@ -74,4 +74,13 @@
             }
         }
     }
+
```
Comment (Zinston) [inline addons/web/static/src/core/notebook/notebook.scss:77]: Mini remarks concerning the commit message (just grammar and stuff, the content is great):

- The "commandbar" is called "Powerbox" ^^'
- "there was functional issues" -> "there _were_ functional issues"
- "responsivenessof" -> "responsiveness of"
- "In a old commit (`<url>`)" -> "In commit [1]" (and then at the end of the commit message, add "[1]: `<url>`")
- "enought" -> "enough"
- "on the all the editor toolbars" -> "on all editor toolbars" (or "on all toolbars of the editor" (@dmo-odoo ! Found one in the wild!)
- "an horizontal scroll" -> "a horizontal scroll"

File: addons/web_editor/static/src/scss/wysiwyg.scss [hunk 22]
```diff
@@ -154,7 +160,9 @@ $o-we-zindex: $o-we-overlay-zindex + 1 !default;
 
 // Toolbar - Generic rules shared among backend & frontend
 .oe-toolbar {
-    display: grid;
+
```
Comment (Zinston) [inline addons/web_editor/static/src/scss/wysiwyg.scss:163]: @stefanorigano knows the guidelines better than I do but I believe this line is superfluous?

File: addons/mass_mailing/static/src/scss/mass_mailing.ui.scss [hunk 23]
```diff
@@ -356,6 +356,14 @@ body.editor_enable.o_basic_theme.o_in_iframe {
                 overflow-y: scroll;
                 max-height: 200px;
             }
+
+            #justify , #style.dropup, #font-size.dropup {
+                > .dropdown-menu {
+                    transform: unset !important;
```
Comment (Zinston) [inline addons/mass_mailing/static/src/scss/mass_mailing.ui.scss:362]: A little comment would be welcome here because without knowing the context, this looks like alarming stuff :-P

File: addons/mass_mailing/static/src/scss/mass_mailing.ui.scss [hunk 24]
```diff
@@ -356,6 +356,14 @@ body.editor_enable.o_basic_theme.o_in_iframe {
                 overflow-y: scroll;
                 max-height: 200px;
             }
+
```
Comment (Zinston) [inline addons/mass_mailing/static/src/scss/mass_mailing.ui.scss:359]: Commit message grammar moment:
- "This commit fix" -> "This commit fixes"

Now about its content, you name the rules that are problematic but you give no information about the reason. I know we don't fully know it, but we at least can say that it's the Popper library that manipulates `transform` and we need to stop it from doing so because its calculations are all wrong.

File: addons/web_editor/static/src/js/editor/toolbar.js [hunk 25]
```diff
@@ -2,6 +2,7 @@ odoo.define('web_editor.toolbar', function (require) {
 'use strict';
 
 var Widget = require('web.Widget');
+var config = require('web.config');
```
Comment (Zinston) [inline addons/web_editor/static/src/js/editor/toolbar.js:5]: I'm gonna seize my opportunity to scold @dmo-odoo here for not writing a commit message. A chance like that is once in a lifetime. So David, consider yourself scolded. :partying_face:

File: addons/web_editor/static/src/xml/editor.xml [hunk 26]
```diff
@@ -9,7 +9,7 @@
         <div id="toolbar" class="oe-toolbar oe-floating">
             <div id="style" class="btn-group dropdown">
                 <button type="button" class="btn dropdown-toggle"
-                    data-bs-toggle="dropdown" data-bs-original-title="Heading" tabindex="-1" aria-expanded="false">
+                    data-bs-toggle="dropdown" data-bs-original-title="Text style" tabindex="-1" aria-expanded="false">
```
Comment (Zinston) [inline addons/web_editor/static/src/xml/editor.xml:12]: You introduced this in your first commit and are changing again here. I suggest you directly name it "Text style" rather than "Heading" in your first commit.

image

File: addons/mass_mailing/static/src/scss/mass_mailing.ui.scss (hunk 24 above)
Comment (Zinston) [reply addons/mass_mailing/static/src/scss/mass_mailing.ui.scss:359]: I see you copy-pasted from my comment directly into your code and into the commit message but I just wrote that to say _what_ you should explain ^^' This doesn't really describe the problem you fixed, doesn't address the `bottom` and `top` properties, and isn't really a full sentence :-P (also, please add a space after a `//`)

File: addons/mass_mailing/static/src/scss/mass_mailing.ui.scss [hunk 27]
```diff
@@ -344,19 +344,26 @@ body.editor_enable.o_basic_theme.o_in_iframe {
...
+                }
+            }
 
-    > .o_we_customize_panel .oe-toolbar {
-        #style.dropup, #font-size.dropup {
-            > .dropdown-menu {
-                overflow-y: scroll;
-                max-height: 200px;
+            // Popper.js manipulates transform and position of the dropdown, and in mass_mailing, its calculations are wrong.
+            // These !important rules stops it from doing so, by unsetting the automatic rules generated by Popper.
```
Comment (Zinston) [inline addons/mass_mailing/static/src/scss/mass_mailing.ui.scss:359]: Thanks, this is a lot better :+1: Only one thing: it should be "stop" instead of "stops" (here and in the commit message) but I'm not going to block the PR for this.

File: addons/web_editor/static/src/js/editor/toolbar.js (hunk 25 above)
Comment (dmo-odoo) [reply addons/web_editor/static/src/js/editor/toolbar.js:5]: I have done so, good sir. 🫥

**Comment by dmo-odoo** [comment]
@robodoo rebase-ff r+

File: addons/web_editor/static/src/xml/editor.xml [hunk 28]
```diff
@@ -112,16 +114,37 @@
...
-                <div id="ordered" data-call="toggleList" data-arg1="OL" title="Toggle ordered list" class="oe-toggle-ordered fa fa-list-ol fa-fw btn"></div>
-                <div id="checklist" data-call="toggleList" data-arg1="CL" title="Toggle checklist" class="oe-toggle-checklist btn fa fa-fw">
-                    <div class="small">
-                        <div class="fa fa-square-o d-block small"/>
-                        <div class="fa fa-check-square d-block small"/>
+            <t t-if="widget.isMobile()"> 
+                <button type="button" class="btn dropdown-toggle"
+                    data-bs-toggle="dropdown" title="Text align" tabindex="-1"
+                    data-bs-original-title="Paragrap" aria-expanded="false">
+                    <i id="paragraphDropdownButton" class="fa fa-list fa-fw"></i>
```
[the remark this answers is not in the document]
Comment (Zinston) [reply addons/web_editor/static/src/xml/editor.xml:121]: I guess that's an oversight. This is wrapped in a `t-if` though so maybe the error is to have the two buttons when the if is true in the first place? If look a bit above, we have a bunch of xml for if not mobile, then we close that if statement, add the `paragraphDropdownButton` (for mobile and not mobile), then open a new if mobile and put that same button again in its mobile version. My guess is that the `</t>` for the original if not mobile should be lower.

File: addons/web_editor/static/src/xml/editor.xml [hunk 29]
```diff
@@ -111,16 +114,37 @@
...
-                <div id="unordered" data-call="toggleList" data-arg1="UL" title="Toggle unordered list" class="oe-toggle-unordered fa fa-list-ul fa-fw btn"></div>
-                <div id="ordered" data-call="toggleList" data-arg1="OL" title="Toggle ordered list" class="oe-toggle-ordered fa fa-list-ol fa-fw btn"></div>
-                <div id="checklist" data-call="toggleList" data-arg1="CL" title="Toggle checklist" class="oe-toggle-checklist btn fa fa-fw">
-                    <div class="small">
-                        <div class="fa fa-square-o d-block small"/>
-                        <div class="fa fa-check-square d-block small"/>
+            <t t-if="widget.isMobile()"> 
+                <button type="button" class="btn dropdown-toggle"
+                    data-bs-toggle="dropdown" title="Text align" tabindex="-1"
+                    data-bs-original-title="Paragrap" aria-expanded="false">
```
Comment (Zinston) [inline addons/web_editor/static/src/xml/editor.xml:120]: Also, @cammarosano while you're at it maybe you can phix the typo :-P

File: addons/web_editor/static/src/xml/editor.xml (hunk 28 above)
[the remark this answers is not in the document]
Comment (cammarosano) [reply addons/web_editor/static/src/xml/editor.xml:121]: I've opened a PR: https://github.com/odoo/odoo/pull/118364
